### PR TITLE
JDK-8332423 : [PPC64] Remove C1_MacroAssembler::call_c_with_frame_resize

### PR DESCRIPTION
**Pull request by suchismith1993**
JBS Issue : [JDK-8332423](https://bugs.openjdk.org/browse/JDK-8332423)
C1_MacroAssembler::call_c_with_frame_resize is only used with frame_resize == 0. 
Also, call_c is adapted as per endianess of system. 
We can adapt the exisiting code to handle the endianness check at one place and not have to repeatedly check at multiple places to make calls to call_c.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332423](https://bugs.openjdk.org/browse/JDK-8332423): [PPC64] Remove C1_MacroAssembler::call_c_with_frame_resize (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Varada M](https://openjdk.org/census#varadam) (@varada1110 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19947/head:pull/19947` \
`$ git checkout pull/19947`

Update a local copy of the PR: \
`$ git checkout pull/19947` \
`$ git pull https://git.openjdk.org/jdk.git pull/19947/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19947`

View PR using the GUI difftool: \
`$ git pr show -t 19947`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19947.diff">https://git.openjdk.org/jdk/pull/19947.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19947#issuecomment-2308923707)

### CODE REVIEW

File: src/hotspot/cpu/ppc/macroAssembler_ppc.cpp [hunk 1]
```diff
@@ -1035,7 +1035,7 @@ address MacroAssembler::call_c_and_return_to_caller(Register r_function_entry) {
   return branch_to(r_function_entry, /*and_link=*/false);
 }
 
-address MacroAssembler::call_c(address function_entry, relocInfo::relocType rt) {
+address MacroAssembler::call_c(address function_entry,relocInfo::relocType) {
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.cpp:1038]: Please restore the original line.

File: src/hotspot/cpu/ppc/macroAssembler_ppc.hpp [hunk 2]
```diff
@@ -359,11 +359,14 @@ class MacroAssembler: public Assembler {
   address call_c(Register function_entry);
   // For tail calls: only branch, don't link, so callee returns to caller of this function.
   address call_c_and_return_to_caller(Register function_entry);
-  address call_c(address function_entry, relocInfo::relocType rt);
+  address call_c(address function_entry, relocInfo::relocType rt = relocInfo::none);
 #else
   // Call a C function via a function descriptor and use full C
   // calling conventions. Updates and returns _last_calls_return_pc.
   address call_c(Register function_descriptor);
+  address call_c(address function_entry, relocInfo::relocType rt = relocInfo::relocType::none) {
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.hpp:367]: I think `relocInfo::none` should be used like above.

File: src/hotspot/cpu/ppc/macroAssembler_ppc.hpp [hunk 3]
```diff
@@ -416,7 +419,6 @@ class MacroAssembler: public Assembler {
   void call_VM_leaf(address entry_point, Register arg_1);
   void call_VM_leaf(address entry_point, Register arg_1, Register arg_2);
   void call_VM_leaf(address entry_point, Register arg_1, Register arg_2, Register arg_3);
-
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.hpp:419]: Please don't remove the empty line.

File: src/hotspot/cpu/ppc/interp_masm_ppc_64.cpp [hunk 4]
```diff
@@ -135,15 +135,7 @@ void InterpreterMacroAssembler::check_and_handle_popframe(Register scratch_reg)
...
-#if defined(ABI_ELFv2)
-    call_c(CAST_FROM_FN_PTR(address,
-                            Interpreter::remove_activation_preserving_args_entry),
-           relocInfo::none);
-#else
-    call_c(CAST_FROM_FN_PTR(FunctionDescriptor*,
-                            Interpreter::remove_activation_preserving_args_entry),
-           relocInfo::none);
-#endif
+    call_c(CAST_FROM_FN_PTR(address, Interpreter::remove_activation_preserving_args_entry), relocInfo::none);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/interp_masm_ppc_64.cpp:138]: `relocInfo::none` can be omitted. It's the default.

File: src/hotspot/cpu/ppc/stubGenerator_ppc.cpp [hunk 5]
```diff
@@ -517,6 +517,106 @@ class StubGenerator: public StubCodeGenerator {
...
+    // Create runtime stub with OopMap.
+    RuntimeStub* stub =
+      RuntimeStub::new_runtime_stub(name, &code,
+                                    /*frame_complete=*/ (int)(frame_complete_pc - start),
+                                    frame_size_in_bytes/wordSize,
+                                    oop_maps,
+                                    false);
+    return stub->entry_point();
+  }
+#undef __
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/stubGenerator_ppc.cpp:619]: I guess this was added unintentionally?

File: src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp [hunk 6]
```diff
@@ -1464,13 +1464,7 @@ address TemplateInterpreterGenerator::generate_native_entry(bool synchronized) {
...
 
   __ mr(R3_ARG1, R16_thread);
-#if defined(ABI_ELFv2)
-  __ call_c(CAST_FROM_FN_PTR(address, JavaThread::check_special_condition_for_native_trans),
-            relocInfo::none);
-#else
-  __ call_c(CAST_FROM_FN_PTR(FunctionDescriptor*, JavaThread::check_special_condition_for_native_trans),
-            relocInfo::none);
-#endif
+  __ call_c(CAST_FROM_FN_PTR(address, JavaThread::check_special_condition_for_native_trans), relocInfo::none);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/templateInterpreterGenerator_ppc.cpp:1467]: `relocInfo::none` can be omitted. It's the default.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
Nice cleanup! I have a couple of small change requests.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. Thanks for the update!

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
One more thing: `C1_MacroAssembler::call_c_with_frame_resize` should be removed completely.

**Review by TheRealMDoerr — APPROVED** [review]
This looks good, now! Thank you!

**Comment by TheRealMDoerr** [comment]
> @TheRealMDoerr i would need another review for this right ?

Yes, please. It's not classified as trivial. Maybe you can ask somebody from your team.

**Comment by varada1110** [comment]
LGTM!
tier1 testing done on linux-ppc64le with both release and fastdebug, no related failures.
Thank you

**Comment by TheRealMDoerr** [comment]
I had already tested on linux-ppc64le. What about AIX?

File: src/hotspot/cpu/ppc/macroAssembler_ppc.hpp [hunk 7]
```diff
@@ -359,11 +359,14 @@ class MacroAssembler: public Assembler {
...
   // For tail calls: only branch, don't link, so callee returns to caller of this function.
   address call_c_and_return_to_caller(Register function_entry);
-  address call_c(address function_entry, relocInfo::relocType rt);
+  address call_c(address function_entry, relocInfo::relocType rt = relocInfo::none);
 #else
   // Call a C function via a function descriptor and use full C
   // calling conventions. Updates and returns _last_calls_return_pc.
   address call_c(Register function_descriptor);
+  address call_c(address function_entry, relocInfo::relocType rt = relocInfo::none) {
+    return call_c((FunctionDescriptor*)function_entry, rt);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.hpp:368]: This breaks ABIv1. Please use a cast to `const FunctionDescriptor*` and move your new function below ` address call_c(const FunctionDescriptor* function_descriptor, relocInfo::relocType rt);`.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/ppc/macroAssembler_ppc.hpp [hunk 8]
```diff
@@ -359,12 +359,16 @@ class MacroAssembler: public Assembler {
...
 #else
   // Call a C function via a function descriptor and use full C
   // calling conventions. Updates and returns _last_calls_return_pc.
   address call_c(Register function_descriptor);
-  // For tail calls: only branch, don't link, so callee returns to caller of this function.
+  address call_c(const FunctionDescriptor* function_descriptor, relocInfo::relocType rt);
+  address call_c(address function_entry, relocInfo::relocType rt = relocInfo::none) {
+    return call_c((const FunctionDescriptor*)function_entry, rt);
+  }
+    // For tail calls: only branch, don't link, so callee returns to caller of this function.
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.hpp]: This looks messed up! I requested to move it down, not do duplicate any function declaration. Also, the comment indentation is bad.

File: src/hotspot/cpu/ppc/macroAssembler_ppc.hpp [hunk 9]
```diff
@@ -359,14 +359,17 @@ class MacroAssembler: public Assembler {
...
   // For tail calls: only branch, don't link, so callee returns to caller of this function.
   address call_c_and_return_to_caller(Register function_entry);
-  address call_c(address function_entry, relocInfo::relocType rt);
+  address call_c(address function_entry, relocInfo::relocType rt = relocInfo::none);
 #else
   // Call a C function via a function descriptor and use full C
   // calling conventions. Updates and returns _last_calls_return_pc.
   address call_c(Register function_descriptor);
-  // For tail calls: only branch, don't link, so callee returns to caller of this function.
+    // For tail calls: only branch, don't link, so callee returns to caller of this function.
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/macroAssembler_ppc.hpp:367]: Please still repair the indentation.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. We should retest it on AIX.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
You effectively changed the type from `relocInfo::runtime_call_type` to `relocInfo::none` in c1_LIRAssembler_ppc.cpp. This is causing problems with ABIv1.
The VM seems to work when switching off C1 on AIX. So, the other files should be ok.

**Comment by suchismith1993** [comment]
> You effectively changed the type from `relocInfo::runtime_call_type` to `relocInfo::none` in c1_LIRAssembler_ppc.cpp. This is causing problems with ABIv1. The VM seems to work when switching off C1 on AIX. So, the other files should be ok.

So do we then call in c1_LIRA using __ call_c(copyfunc_addr, relocInfo::runtime_call_type); ?

**Comment by TheRealMDoerr** [comment]
> > You effectively changed the type from `relocInfo::runtime_call_type` to `relocInfo::none` in c1_LIRAssembler_ppc.cpp. This is causing problems with ABIv1. The VM seems to work when switching off C1 on AIX. So, the other files should be ok.
> 
> So do we then call in c1_LIRA using __ call_c(copyfunc_addr, relocInfo::runtime_call_type); ?

Yes, for all 4 `call_c` in c1_LIRAssembler_ppc.cpp.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. Hopefully the tests will pass on AIX.

**Comment by TheRealMDoerr** [comment]
@varada1110: Can you use "Approve" instead of "Comment", please? Otherwise, your review doesn't get recorded properly.

**Comment by TheRealMDoerr** [comment]
Test results look good this time. I think it's good to go.

**Comment by suchismith1993** [comment]
> Test results look good this time. I think it's good to go.

Hi Martin, you mean the results for AIX ?

**Review by varada1110 — APPROVED** [review]
LGTM! Thank you Martin for testing.

**Comment by suchismith1993** [comment]
thank you all for the review and tests.

**Pull request closed without merging**